src/package/providers/nix.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright 2015-2017 Intecture Developers. See the COPYRIGHT file at the
+// top-level directory of this distribution and at
+// https://intecture.io/COPYRIGHT.
+//
+// Licensed under the Mozilla Public License 2.0 <LICENSE or
+// https://www.tldrlegal.com/l/mpl-2.0>. This file may not be copied,
+// modified, or distributed except according to those terms.
+
+//! Nix package provider
+
+use command::{Command, CommandResult};
+use error::Result;
+use host::Host;
+use super::*;
+
+pub struct Nix;
+
+impl Provider for Nix {
+    fn get_providers(&self) -> Providers {
+        Providers::Nix
+    }
+
+    fn is_active(&self, host: &mut Host) -> Result<bool> {
+        let cmd = Command::new("which nix-env");
+        let result = try!(cmd.exec(host));
+
+        Ok(result.exit_code == 0)
+    }
+
+    fn is_installed(&self, _: &mut Host, _: &str) -> Result<bool> {
+        // Nix packages can be installed by name, attribute path
+        // or full Nix store path. They are not all queried the same
+        // way. Also querying by name is very slow. Considering that
+        // the Nix package manager is atomic and that it doesn't
+        // reinstall a package that's already on the system, it's
+        // better to just assume that the package is not installed
+        // and try to install it when needed.
+        Ok(false)
```

Comment:
> While NixOS may have an atomic package manager, other Intecture code may depend on reacting to a package either existing or not existing.

I thought that `is_installed` would only be called to decide whether to run `install` or `uninstall`. Do we (or do we plan to) use it for anything else?

> Plus it's totally misleading to the user and will almost certainly cause consternation. We need to implement it correctly.

Due to Nix's nature, querying a Nix package reliably takes a lot of effort to implement. There are a lot of considerations to be made. For example:-

* Is the package being queried by name, attribute path, nix store path, by file system path or by URL (a Nix package can be installed using any of these ways).
* Nix packages can either be installed declaratively (using config files) or imperatively (using `nix-env`). Packages installed declaratively are not seen by `nix-env --query` so we would have to first resolve the package to a nix store path and check if it exists in the Nix store.
* Even if a package is in the nix store, is it not just waiting to be garbage collected? If so we would have to re-register it so it's not garbage collected.

Now that I think about it, querying a Nix package's status for our use case might actually warrant its own crate. I would rather just document the fact that `is_installed` always returns `false` for Nix packages for now. Another option might be to return an `Error::UnsupportedOperation` or something. How does `Intecture` handle errors from this method?

> Also, speed isn't a huge issue at the moment. There won't be a large volume of transactions - Intecture code is only run sporadically. How long would a name search take?

I haven't actually tried to time this but from experience it can take an insanely huge amount, like say 10 minutes.